text/src/swam/witx/parser/TypesParser.scala

```diff
@@ -0,0 +1,102 @@
+package swam
+package witx
+package parser
+
+import fastparse._
+import swam.witx.parser.parser.WitxWhitespace._
+import swam.witx.unresolved._
+
+import scala.collection.MapView
+
+/**
+    @author Javier Cabrera-Arteaga on 2020-03-18
+  */
+object TypesParser {
+
+  import Lexical._
+
+  private var declaredTypes = Map[String, BaseWitxType](
+    "u32" -> BasicType("u32"),
+    "u64" -> BasicType("u64"),
+    "u8" -> BasicType("u8"),
+    "u16" -> BasicType("u16"),
+    "string" -> BasicType("string")
+  )
+
+  def file[_: P]: P[Map[String, BaseWitxType]] = {
+    P(ws ~ types ~ ws ~ End).map(_ => declaredTypes)
+  }
+
+  def types[_: P] = {
+    P("(" ~ word("typename") ~ id ~/ subtype ~ ")")
+      .map {
+        case (name, tp) => {
+          declaredTypes = declaredTypes + (name -> tp)
+        }
+      }
+      .rep(1)
+  }
+
+  def subtype[_: P]: P[BaseWitxType] = {
+    P(
+      name | enum | flags | struct | array | handle | union
+    ).map {
+      case simpleType: String        => AliasType(simpleType)
```

Comment:
This is really bad practice, `String` and `BaseWitxType` have no common super type, meaning that you defeat the all static type checker here. In the `name` alternative, map directly to `AliasType` and you are done.

Reply:
Gotcha